cc/algorithms/partition-selection-strategies.h

```
@@ -0,0 +1,345 @@
+  }
+
+ protected:
+  double adjusted_epsilon_;
+  double crossover_1_;
+  double crossover_2_;
+
+  double ProbabilityOfKeep(double n) {
+    if(n == 0) {
+      return 0;
+    }
+    else if (n <= crossover_1_) {
+      return (((exp(n * adjusted_epsilon_) - 1) / (exp(adjusted_epsilon_) - 1))
+              * adjusted_delta_);
+    }
+    else if (n > crossover_1_ && n <= crossover_2_) {
+      double m = n - crossover_1_;
+      return ((1 - exp(-1 * m * adjusted_epsilon_))
+               * (1 + adjusted_delta_ / (exp(adjusted_epsilon_) - 1))
+               + exp(-1 * m * adjusted_epsilon_)
+               * ProbabilityOfKeep(crossover_1_));
+    }
+    else {
+      return 1;
+    }
+  }
+
+};
+
+
+class LaplacePartitionSelection : public PartitionSelectionStrategy {
+ public:
+  //Builder for LaplacePartitionSelection
+  class Builder : public PartitionSelectionStrategy::Builder {
+   public:
+     Builder& SetLaplaceMechanism(LaplaceMechanism::Builder* laplace_builder) {
+       laplace_builder_ = laplace_builder;
+       return *static_cast<Builder*>(this);
+     }
+
+     base::StatusOr<std::unique_ptr<PartitionSelectionStrategy>>
+     Build() override {
+       base::Status max_partitions_contributed_status =
+         MaxPartitionsContributedIsSetAndValid();
+       base::Status delta_status = DeltaIsSetAndValid();
+       if (!max_partitions_contributed_status.ok()) {
+         return max_partitions_contributed_status;
+       }
+       else if (!epsilon_.has_value()) {
+         return base::InvalidArgumentError("Epsilon has to be set.");
+       }
+       else if (!delta_status.ok()) {
+         return delta_status;
+       }
+       else if (laplace_builder_.has_value()) {
+         std::unique_ptr<NumericalMechanism> mechanism_ =
+           (*(laplace_builder_.value())).SetEpsilon(epsilon_.value())
+           .SetL1Sensitivity(max_partitions_contributed_.value()).Build()
+           .ValueOrDie();
```

ValueOrDie is rather dangerous - the code will crash if there was an error. You should return any errors instead (see the ASSIGN_OR_RETURN macro in status_macros.h).

---

`ValueOrDie` has been recently marked deprecated.  You can use `ASSIGN_OR_RETURN`.